tests/test_direct_solver_device.cu

```diff
@@ -0,0 +1,117 @@
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#define BOOST_TEST_MODULE direct_solver_device
+
+#include "main.cc"
+
+#include <mfmg/dealii_operator_device.cuh>
+#include <mfmg/exceptions.hpp>
+#include <mfmg/utils.cuh>
+
+#include <deal.II/lac/cuda_vector.h>
+#include <deal.II/lac/read_write_vector.h>
+#include <deal.II/lac/sparse_direct.h>
+
+BOOST_AUTO_TEST_CASE(direct_solver)
+{
+  // Create the cusolver_dn_handle
+  cusolverDnHandle_t cusolver_dn_handle = nullptr;
+  cusolverStatus_t cusolver_error_code;
+  cusolver_error_code = cusolverDnCreate(&cusolver_dn_handle);
+  mfmg::ASSERT_CUSOLVER(cusolver_error_code);
+  // Create the cusolver_sp_handle
+  cusolverSpHandle_t cusolver_sp_handle = nullptr;
+  cusolver_error_code = cusolverSpCreate(&cusolver_sp_handle);
+  mfmg::ASSERT_CUSOLVER(cusolver_error_code);
+  // Create the cusparse_handle
+  cusparseHandle_t cusparse_handle = nullptr;
+  cusparseStatus_t cusparse_error_code;
+  cusparse_error_code = cusparseCreate(&cusparse_handle);
+  mfmg::ASSERT_CUSPARSE(cusparse_error_code);
+
+  // Create the matrix on the host.
+  dealii::SparsityPattern sparsity_pattern;
+  dealii::SparseMatrix<double> matrix;
+  unsigned int const size = 30;
+  std::vector<std::vector<unsigned int>> column_indices(size);
+  for (unsigned int i = 0; i < size; ++i)
+  {
+    unsigned int j_max = std::min(size, i + 2);
+    unsigned int j_min = (i == 0) ? 0 : i - 1;
+    for (unsigned int j = j_min; j < j_max; ++j)
+      column_indices[i].emplace_back(j);
+  }
+  sparsity_pattern.copy_from(size, size, column_indices.begin(),
+                             column_indices.end());
+  matrix.reinit(sparsity_pattern);
+  for (unsigned int i = 0; i < size; ++i)
+  {
+    unsigned int j_max = std::min(size - 1, i + 1);
+    unsigned int j_min = (i == 0) ? 0 : i - 1;
+    matrix.set(i, j_min, -1.);
+    matrix.set(i, j_max, -1.);
+    matrix.set(i, i, 4.);
+  }
+  // Solve on the host.
+  dealii::SparseDirectUMFPACK direct_solver;
+  direct_solver.initialize(matrix);
+  dealii::Vector<double> sol_ref(size);
+  for (auto &val : sol_ref)
+    val = 1.;
+  direct_solver.solve(sol_ref);
+
+  // Move the matrix and the rhs to the host
+  mfmg::SparseMatrixDevice<double> matrix_dev(mfmg::convert_matrix(matrix));
+  matrix_dev.cusparse_handle = cusparse_handle;
+  cusparse_error_code = cusparseCreateMatDescr(&matrix_dev.descr);
+  mfmg::ASSERT_CUSPARSE(cusparse_error_code);
+  cusparse_error_code =
+      cusparseSetMatType(matrix_dev.descr, CUSPARSE_MATRIX_TYPE_GENERAL);
+  mfmg::ASSERT_CUSPARSE(cusparse_error_code);
+  cusparse_error_code =
+      cusparseSetMatIndexBase(matrix_dev.descr, CUSPARSE_INDEX_BASE_ZERO);
+  mfmg::ASSERT_CUSPARSE(cusparse_error_code);
+  dealii::LinearAlgebra::ReadWriteVector<double> rw_vector(size);
```

Comment:
I'm not sure I understand. Do you use `ReadWriteVector` to simplify the code to copy on the device? Looking at the documentation, I don't see anything about the device there...

Reply:
Yeah `ReadWriteVector` lives on the host and `CUDAWrappers::Vector` lives on the device. To copy to the host, I do
```
rw_vector.import(cuda_vector, dealii::VectorOperation::insert);
```
and to copy to the host, I just do
```
cuda_vector.import(rw_vector, dealii::VectorOperation::insert);
```

Reply:
So what structure is device aware? Is it `import` call, or `insert`, or something else?

Reply:
I am rewriting the test.